File: tools/binding_compliance/conformance/families/hash_cache_controls.py
"""Actual empty-state cache controls; populated-state tests remain independent."""

from collections.abc import Mapping
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy


def _observed(observation: Mapping[str, Any]) -> bool:
    """Require complete public zero-state statistics and independently returned sizes."""
    if set(observation) != {
        "initial",
        "initialSize",
        "afterReset",
        "afterClear",
        "finalSize",
    }:
        return False
    empty = {"hits": 0, "misses": 0, "hitRate": "0.000", "size": 0, "capacity": 1024}
    return (
        all(
            observation[key] == empty for key in ("initial", "afterReset", "afterClear")
        )
        and type(observation["initialSize"]) is int
        and observation["initialSize"] == 0
        and type(observation["finalSize"]) is int
        and observation["finalSize"] == 0
    )


HASH_CACHE_CONTROLS_COVERAGE_POLICY = FamilyCoveragePolicy(
    "hash-cache-controls",
    (
        CoveragePredicate(
            "empty-cache-controls",
            "hash-cache-controls.empty",
            "hash-cache-controls.empty",
            "cache-state",
            ("clear_cache", "cache_size", "cache_stats", "reset_cache_stats"),
            _observed,
            runtime_operations=(
                "hash_cache_clear",
                "hash_cache_size",
                "hash_cache_stats",
                "reset_hash_cache_stats",
            ),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/hash_cache_controls.rs
//! Empty-cache control observations for the public CXX-accessible state.

use super::{RunnerResult, invalid};
use classic_file_io_core::hash::FileHasher;
use serde_json::{Value, json};

/// Projects every portable statistic using an explicit canonical decimal ratio.
fn stats() -> Value {
    let value = FileHasher::cache_stats();
    json!({"hits":value.hits,"misses":value.misses,"hitRate":format!("{:.3}",value.hit_rate),"size":value.size,"capacity":value.capacity})
}

/// Exercises idempotent empty-state control calls without claiming populated-cache execution.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    if fixture["operation"] != "empty-controls" {
        return Err(invalid("unsupported cache control fixture").into());
    }
    FileHasher::clear_cache();
    FileHasher::reset_cache_stats();
    let initial = stats();
    let initial_size = FileHasher::cache_size();
    FileHasher::reset_cache_stats();
    let reset = stats();
    FileHasher::clear_cache();
    Ok(
        json!({"initial":initial,"initialSize":initial_size,"afterReset":reset,"afterClear":stats(),"finalSize":FileHasher::cache_size()}),
    )
}
